lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -517,6 +833,22 @@ CanonicalizeInstruction::canonicalize(SILInstruction *inst) {
   if (auto nextII = simplifyAndReplace(inst, *this))
     return nextII.getValue();
 
+  if (auto *tuple = dyn_cast<TupleInst>(inst)) {
+    if (auto tupleTy = tuple->getType().getAs<TupleType>()) {
+      if (!tupleTy->hasAnyNamedElement()) {
```

Comment:
Out of curiosity: what's the problem with named tuple elements?

Reply:
@eeckstein as an example, consider the following SIL:

```
// thunk for @callee_guaranteed (@unowned KeyPathBuffer) -> (@unowned UnsafeMutablePointer<B>, @owned Swift.AnyObject?, @error @owned Error)
sil shared [transparent] [serializable] [reabstraction_thunk] [ossa] @$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR : $@convention(thin) <Root, Value> (KeyPathBuffer, @noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)) -> (@out (pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), @error Error) {
// %0                                             // users: %4, %3
// %1                                             // user: %5
// %2                                             // user: %5
bb0(%0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), %1 : $KeyPathBuffer, %2 : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)):
  %3 = tuple_element_addr %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), 0 // user: %9
  %4 = tuple_element_addr %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), 1 // user: %10
  try_apply %2(%1) : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error), normal bb1, error bb2 // id: %5

// %6                                             // user: %7
bb1(%6 : @owned $(UnsafeMutablePointer<Value>, Optional<AnyObject>)): // Preds: bb0
  (%7, %8) = destructure_tuple %6 : $(UnsafeMutablePointer<Value>, Optional<AnyObject>) // users: %9, %10
  store %7 to [trivial] %3 : $*UnsafeMutablePointer<Value> // id: %9
  store %8 to [init] %4 : $*Optional<AnyObject>   // id: %10
  %11 = tuple ()                                  // user: %12
  return %11 : $()                                // id: %12

// %13                                            // user: %14
bb2(%13 : @owned $Error):                         // Preds: bb0
  throw %13 : $Error                              // id: %14
} // end sil function '$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR'
```

This we (before some changes I am going to make real quick) we would transform to:

```
sil shared [transparent] [serializable] [reabstraction_thunk] [ossa] @$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR : $@convention(thin) <Root, Value> (KeyPathBuffer, @noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)) -> (@out (pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), @error Error) {
// %0                                             // user: %5
// %1                                             // user: %3
// %2                                             // user: %3
bb0(%0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>), %1 : $KeyPathBuffer, %2 : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error)):
  try_apply %2(%1) : $@noescape @callee_guaranteed (KeyPathBuffer) -> (UnsafeMutablePointer<Value>, @owned Optional<AnyObject>, @error Error), normal bb1, error bb2 // id: %3

// %4                                             // user: %5
bb1(%4 : @owned $(UnsafeMutablePointer<Value>, Optional<AnyObject>)): // Preds: bb0
  store %4 to [init] %0 : $*(pointer: UnsafeMutablePointer<Value>, owner: Optional<AnyObject>) // id: %5
  %6 = tuple ()                                   // user: %7
  return %6 : $()                                 // id: %7

// %8                                             // user: %9
bb2(%8 : @owned $Error):                          // Preds: bb0
  throw %8 : $Error                               // id: %9
} // end sil function '$ss13KeyPathBufferVSpyq_GyXlSgs5Error_pIgydozo_AbC7pointer_AD5ownertsAE_pIegyrzo_r0_lTR'
```

The verifier throws since the types do not match up. We really shouldn't care about tuple element names at the SIL level and should in type lowering eliminate them IMO. But that is more than I want to do in this PR (which is meant to only solve problems exposed by homogeneous tuples).

Reply:
You are right. I found a way to do it by checking if the types match. Thanks!